subsys/bluetooth/services/bas.c

```diff
@@ -0,0 +1,143 @@
+
+#include <sensor.h>
+
+#define LOG_MODULE_NAME bas
+#define LOG_LEVEL CONFIG_BT_GATT_BAS_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER();
+
+static struct bt_gatt_service bas;
+static struct bt_gatt_attr attrs[];
+static struct bt_gatt_ccc_cfg charge_lvl_ccc_cfg[BT_GATT_CCC_MAX];
+
+static struct device *battery;
+static struct sensor_value charge_lvl;
+static bool notif_enabled;
+
+int bas_notify(void);
+
+static void charge_lvl_ccc_cfg_changed(const struct bt_gatt_attr *attr,
+				       u16_t value)
+{
+	ARG_UNUSED(attr);
+
+	notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+	LOG_INF("Notifications %s", notif_enabled ? "enabled" : "disabled");
+}
+
+static void charge_lvl_fetch(void)
+{
+	int rc;
+
+	__ASSERT(battery != NULL, "No binding to battery sensor!\n");
+
+	rc = sensor_sample_fetch_chan(battery, SENSOR_CHAN_STATE_OF_CHARGE);
+	if (rc) {
+		LOG_ERR("Could not obtain new sensor reading, err %d", rc);
+		return;
+	}
+
+	rc = sensor_channel_get(battery, SENSOR_CHAN_STATE_OF_CHARGE,
+				&charge_lvl);
+	if (rc) {
+		LOG_ERR("Could not read sensor data, err %d", rc);
+		return;
+	}
+}
+
+static ssize_t charge_lvl_read(struct bt_conn *conn,
+			       const struct bt_gatt_attr *attr, void *buf,
+			       u16_t len, u16_t offset)
+{
+	u8_t lvl8 = (u8_t)(charge_lvl.val1);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &lvl8,
+				 sizeof(lvl8));
+}
+
+void level_changed_callback(struct device *dev, struct sensor_trigger *trigger)
+{
+	ARG_UNUSED(dev);
+
+	if (trigger->type == SENSOR_TRIG_DATA_READY &&
+	    trigger->chan == SENSOR_CHAN_STATE_OF_CHARGE) {
+		/* fetch battery level and send notifications, if enabled */
+		charge_lvl_fetch();
+		bas_notify();
+	}
+}
+
+static int bas_init(struct device *dev)
+{
+	int rc;
+
+	ARG_UNUSED(dev);
+
+	battery = device_get_binding(CONFIG_BATTERY_SENSOR_DEV_NAME);
+	if (!battery) {
+		LOG_ERR("Could not bind to device %s",
+			CONFIG_BATTERY_SENSOR_DEV_NAME);
+		return -EIO;
+	}
+
+	struct sensor_trigger trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_STATE_OF_CHARGE
+	};
+
+	rc = sensor_trigger_set(battery, &trigger, level_changed_callback);
+	if (rc) {
+		LOG_ERR("Could not set up sensor trigger, err %d", rc);
+		return -EIO;
+	}
+
+	return bt_gatt_service_register(&bas);
+}
+
+int bas_notify(void)
+{
+	u8_t lvl8;
+
+	if (!notif_enabled) {
+		return -EINVAL;
+	}
+
+	lvl8 = (u8_t)(charge_lvl.val1);
+
+	return bt_gatt_notify(NULL, &attrs[1], &lvl8, sizeof(lvl8));
+}
+
+/* Battery Service Declaration */
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_BAS),
+	BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL,
+		BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+		BT_GATT_PERM_READ, charge_lvl_read, NULL, &charge_lvl),
+	BT_GATT_CCC(charge_lvl_ccc_cfg, charge_lvl_ccc_cfg_changed),
+};
+
+static struct bt_gatt_service bas = BT_GATT_SERVICE(attrs);
+
+SYS_INIT(bas_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
please don't do it.... I would appreciate it application could initialize bas on its own when it is ready

Reply:
This just registers the service in the database, if you have this service that's probably what you want to do. This has been already discussed at lengths in this PR and #9712.